src/CombFilter/CombFilter.cpp

```diff
@@ -0,0 +1,93 @@
+//
+// Created by Rose Sun on 2/12/22.
+//
+
+#include "CombFilter.h"
+#include "RingBuffer.h"
+#include "ErrorDef.h"
+
+CCombFilterBase::CCombFilterBase (float fMaxDelayLengthInS, float fSampleRateInHz, int iNumChannels) :
+fMaxDelayLengthInS(fMaxDelayLengthInS),
+fSampleRateInHz(fSampleRateInHz),
+fDelayLengthInS(0.f),
+iNumChannels(iNumChannels)
+{
+    this->reset ();
+    bufferLength = static_cast<long>(round(fMaxDelayLengthInS * fSampleRateInHz));
+
+    ppRingBuffer = new CRingBuffer<float>*[iNumChannels];
+    for (int i=0; i<iNumChannels; ++i)
+        ppRingBuffer[i] = new CRingBuffer<float>(bufferLength);
+}
+
+Error_t CCombFilterBase::setParam (CCombFilterIf::FilterParam_t eParam, float fParamValue)
+{
+    switch (eParam) {
+        case CCombFilterIf::FilterParam_t::kParamDelay:
+            if (fParamValue > fMaxDelayLengthInS)
+                return Error_t::kFunctionInvalidArgsError;
+            fDelayLengthInS = fParamValue;
+            delayLength = static_cast<long>(round(fSampleRateInHz * fParamValue));
+            for (int i=0; i < iNumChannels; ++i)
+                ppRingBuffer[i]->setWriteIdx(1.F*delayLength + ppRingBuffer[i]->getReadIdx());
+            break;
+        case CCombFilterIf::FilterParam_t::kParamGain:
+            if (fGain < 0)
+                return Error_t::kFunctionInvalidArgsError;
+            fGain = fParamValue;
+            break;
+    }
+    return Error_t::kNoError;
+}
+
+float CCombFilterBase::getParam (CCombFilterIf::FilterParam_t eParam) const
+{
+    switch (eParam) {
+        case CCombFilterIf::FilterParam_t::kParamDelay:
+            return fDelayLengthInS;
+
+        case CCombFilterIf::FilterParam_t::kParamGain:
+            return fGain;
+    }
+    return -1.f;
+}
+
+CCombFilterBase::~CCombFilterBase ()
+{
+    this->reset ();
+}
+
+void CCombFilterBase::reset() {
```

Comment:
I feel maybe the deletion of the RingBuffer is better suited in a `destroy()` method. The `reset()` method should ideally be used to set all existing values of the parameters such that the object can be reused.